Lib/functools.py

```diff
@@ -274,43 +275,129 @@ def reduce(function, sequence, initial=_initial_missing):
 ### partial() argument application
 ################################################################################
 
-# Purely functional, no descriptor behaviour
-class partial:
-    """New function with partial application of the given arguments
-    and keywords.
+
+class _PlaceholderType:
+    """The type of the Placeholder singleton.
+
+    Used as a placeholder for partial arguments.
     """
+    __instance = None
+    __slots__ = ()
 
-    __slots__ = "func", "args", "keywords", "__dict__", "__weakref__"
+    def __init_subclass__(cls, *args, **kwargs):
+        raise TypeError(f"type '{cls.__name__}' is not an acceptable base type")
 
-    def __new__(cls, func, /, *args, **keywords):
+    def __new__(cls):
+        if cls.__instance is None:
+            cls.__instance = object.__new__(cls)
+        return cls.__instance
```

Comment:
This was to mimic C implementation, which has the behaviour of `None`.

I think the question is: "Is it a good practice for a non-trivial sentinel to be singleton, i.e. `type(None)() is None`.

If yes and this sentinel is considered non-trivial, then this is as good as it can get for now and protection issues can be sorted out together with further developments in this area.

If no, then this needs to be changed for both C and Python.

@rhettinger has suggested this initially and I like this behaviour (and adapted it to my own sentinels). It would be good if you together could come to a decision before I make any further changes here.

Reply:
This is not required for its main function, and this complicates both implementations. It is better to implement only necessary parts. If later we will find a need of this feature, it will be easier to add it than to remove it.

Strictly speaking, making the Placeholder class non-inheritable and non-instantiable is not required. But it is easy to implement.

I hope Raymond will change his opinion on this.

Reply:
What you are saying makes sense, but at the same time I like current behaviour and if sentinels were standardised and their creation was made more convenient I think this is nice.

I am neutral by now. Well, slightly negative just because I put thought effort into this and simply like it.